panther_gpiod/src/gpio_driver.cpp

```diff
@@ -0,0 +1,274 @@
+  } catch (const std::out_of_range & err) {
+    throw std::runtime_error("No name defined for one of pins: " + std::string(err.what()));
+  }
+
+  gpiod::line::offset offset = chip.get_line_offset_from_name(pin_name);
+
+  builder.add_line_settings(offset, settings);
+  gpio_info.offset = offset;
+}
+
+gpiod::line_settings GPIODriver::generate_line_settings(const GPIOInfo & gpio_info)
+{
+  auto settings = gpiod::line_settings();
+  settings.set_direction(gpio_info.direction);
+  settings.set_output_value(gpio_info.init_value);
+  settings.set_active_low(gpio_info.active_low);
+
+  if (gpio_info.direction == gpiod::line::direction::INPUT) {
+    settings.set_edge_detection(gpiod::line::edge::BOTH);
+    settings.set_debounce_period(std::chrono::milliseconds(gpio_debounce_period_));
+  }
+
+  return settings;
+}
+
+void GPIODriver::change_pin_direction(const GPIOPin pin, const gpiod::line::direction direction)
+{
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  GPIOInfo & gpio_info = get_gpio_info_ref(pin);
+
+  if (gpio_info.direction == direction) {
+    return;
+  }
+
+  gpio_info.direction = direction;
+
+  auto line_config = gpiod::line_config();
+
+  for (const auto & gpio_info : gpio_info_storage_) {
+    gpiod::line_settings settings = generate_line_settings(gpio_info);
+    line_config.add_line_settings(gpio_info.offset, settings);
+  }
+
+  line_request_->reconfigure_lines(line_config);
+  gpio_info.value = line_request_->get_value(gpio_info.offset);
+}
+
+bool GPIODriver::is_pin_active(const GPIOPin pin)
+{
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  return get_gpio_info_ref(pin).value == gpiod::line::value::ACTIVE;
+}
+
+bool GPIODriver::set_pin_value(const GPIOPin pin, const bool value)
+{
+  GPIOInfo & gpio_info = get_gpio_info_ref(pin);
+
+  if (gpio_info.direction == gpiod::line::direction::INPUT) {
+    throw std::invalid_argument("Cannot set value for INPUT pin.");
+  }
+
+  gpiod::line::value gpio_value = value ? gpiod::line::value::ACTIVE : gpiod::line::value::INACTIVE;
+
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  try {
+    line_request_->set_value(gpio_info.offset, gpio_value);
+
+    if (line_request_->get_value(gpio_info.offset) != gpio_value) {
+      throw std::runtime_error("Failed to change GPIO state.");
+    }
+
+    gpio_info.value = gpio_value;
+    gpio_edge_event_callback(gpio_info);
+
+    return true;
+  } catch (const std::exception & e) {
+    std::cerr << "Error while setting GPIO pin value: " << e.what() << std::endl;
+    return false;
+  }
+}
+
+void GPIODriver::configure_edge_event_callback(
+  const std::function<void(const GPIOInfo &)> & callback)
+{
+  gpio_edge_event_callback = callback;
+}
+
+void GPIODriver::gpio_monitor_on()
+{
+  if (is_gpio_monitor_thread_running()) {
+    return;
+  }
+
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  for (auto & info : gpio_info_storage_) {
+    info.value = line_request_->get_value(info.offset);
+  }
+  lock.unlock();
```

Comment:
```suggestion
  {
    std::unique_lock lock(gpio_info_storage_mutex_);
    for (auto & info : gpio_info_storage_) {
      info.value = line_request_->get_value(info.offset);
    }
  }
```
I prefer to do this with scopes (I think that it is easier to forget about unlock)